Add support for 'flattened object' fields.

jtibshirani opened the pull request:
This PR merges the `object-fields` feature branch. All commits on the branch have been individually code reviewed as part of earlier PRs.

Before merging, there are a few open issues to resolve:
- [x] The field type is currently marked 'experimental'. I've started an internal discussion to see if we can remove this tag, since the feature is useful in its current form and we don't expect huge changes in its API.
- [x] There is some chance we want to revise the type name again -- I reopened an internal issue to ask for feedback.
- [x] I will push a commit with some tweaks to the documentation (I've gained more insight on the field type from performance profiling and discussions with the Kibana team).

Original issue: #25312
Meta-issue tracking design + implementation: #33003

jtibshirani committed:
Add a simple JSON field mapper. (#33923)

* Add a simple JSON field type.
* Add support for ignore_above.
* Add support for null_value.
* Add support for split_queries_on_whitespace.
* Prevent norms from being enabled.
* Clarify the message around copy_to not being supported.
* Disallow wildcard queries.
* For now, disallow the field from being stored.

jtibshirani committed:
When parsing JSON fields, also create tokens prefixed with the field key. (#34207)

jtibshirani committed:
Add support for querying JSON fields based on key. (#34621)

jtibshirani committed:
Add support for storing JSON fields. (#34942)

jtibshirani committed:
Enforce a limit on the depth of the JSON object. (#35063)

jtibshirani committed:
Disallow doc_values in the JSON field mapping. (#35282)

jtibshirani committed:
Make sure stored JSON fields are properly decoded. (#35279)

jtibshirani committed:
Add tests for the supported query types. (#35319)

* Add tests for the supported query types.
* Disallow unbounded range queries on keyed JSON fields.
* Make sure MappedFieldType#hasDocValues always returns false.

jtibshirani committed:
Add documentation for JSON fields. (#35281)

* Add documentation for JSON fields.

jtibshirani committed:
Add a test around JSON fields and source filtering. (#35399)

jtibshirani committed:
Prevent slow field lookups when JSON fields are present. (#39872)

We now track the maximum depth of any JSON field, which allows the JSON field
lookup to be short-circuited as soon as that depth is reached. This helps
prevent slow lookups when the user is searching over a very deep field that is
not in the mappings.

jtibshirani committed:
Add doc values support for JSON fields. (#40069)

When `doc_values` are enabled, we now add two `SortedSetDocValuesFields` for each token: one containing the raw `value`, and another with `key\0value`. The root JSON field uses the standard `SortedSetDVOrdinalsIndexFieldData`. For keyed fields, this PR introduces a new type ` KeyedJsonIndexFieldData` that wraps the standard ordinals field data and filters out values that do not match the right prefix. This gives support for sorting on JSON fields, as well as simple keyword-style aggregations like `terms`.

One slightly tricky aspect is caching of these doc values. Given a keyed JSON field, we need to make sure we don't store values filtered on a certain prefix under the same cache key as ones filtered on a different prefix. However, we also want to load and cache global ordinals only once per keyed JSON field, as opposed to having a separate cache entry per prefix.

jtibshirani committed:
Rename 'json' to 'embedded_json'. (#40712)

One concern around the name `json` is that because the entire document is JSON,
new users may see this field and think that they should always use it. We
thought that a more verbose name like `embedded_json` would help convey that the
field type has a special, non-obvious purpose.

This commit updates documentation references to `embedded_json`, but leaves the
`JsonField` naming in the code to avoid very long class names.

jtibshirani committed:
Rename embedded_json.asciidoc for consistency.

jtibshirani committed:
Rebase keyed JSON ordinals to start from zero. (#41282)

This PR updates `KeyedJsonAtomicFieldData` to always return ordinals in the
range `[0, (maxOrd - minOrd)]`, which is necessary for certain aggregations and
sorting options to be supported.

As discussed in #41220, I opted not to support
`KeyedIndexFieldData#getOrdinalMap`, as it would add substantial complexity.
The one place this affects is the 'low cardinality' optimization for terms
aggregations, which now needs to be disabled for keyed JSON fields.

It was fairly difficult to incorporate this change, and I have a couple
follow-up refactors in mind to help simplify the global ordinals code. (I will
likely wait until this feature branch is merged though before opening PRs on
master).

jtibshirani committed:
Allow non-exact query types on the root JSON field. (#41290)

In an earlier iteration of the design, it made sense to disallow these query
types on the root JSON field. It should now it be fine to allow them.

jtibshirani committed:
Make sure to return keyed JSON field types in FieldTypeLookup#iterator. (#41319)

The index warmer iterates through all field types when determining the fields
for which global ordinals should be loaded. Previously, keyed JSON field types
were not returned from FieldTypeLookup#iterator, so their eager_global_ordinals
setting would be ignored. This PR fixes the issue by including keyed JSON fields
in FieldTypeLookup#iterator.

jtibshirani committed:
Ensure all JSON field types are available for lookup. (#41914)

Previously, if multiple `embedded_json` fields were added at once, only the last
one would be registered with `FieldTypeLookup`. This bug was uncovered when
trying out different scenarios for performance benchmarking.

elasticmachine commented:
Pinging @elastic/es-search

jpountz reviewed (commented):
I really like how queries and aggregations work as if fields had been mapped on their own. However, this is not the case for stored fields, which makes me wonder whether we should leave it unsupported for now.

jtibshirani commented on docs/reference/mapping/types.asciidoc:
```diff
@@ -82,6 +84,8 @@ include::types/date.asciidoc[]
 include::types/date_nanos.asciidoc[]
 
+include::types/embedded-json.asciidoc[]
+
```
👍

jtibshirani replied:
I think these includes are just alphabetized. For the actual links to individual field types, I put it under "Specialised datatypes" to encourage users to think through whether it's appropriate for their data.

jtibshirani commented on docs/reference/mapping/types/embedded-json.asciidoc:
```diff
@@ -0,0 +1,212 @@
+- Only one field mapping is created for the whole object, which can help
+  prevent a <<mapping-limit-settings, mappings explosion>> due to a large
+  number of field mappings.
+- An embedded JSON field may take up less space in the index, as only one underlying
```
+1 I actually have a TODO to rework this docs page, I will ping you for another look when that's done.

jtibshirani commented on server/src/main/java/org/elasticsearch/index/mapper/FieldTypeLookup.java:
```diff
@@ -83,14 +100,89 @@ public FieldTypeLookup copyAndAddAll(String type,
+        for (int i = 0; i < field.length(); ++i) {
+            if (field.charAt(i) == '.') {
+                numDots++;
+            }
```
👍

jtibshirani committed:
Remove the 'experimental' tag.

jtibshirani commented on server/src/main/java/org/elasticsearch/index/mapper/JsonFieldMapper.java:
```diff
@@ -0,0 +1,639 @@
+
+    public static final String CONTENT_TYPE = "embedded_json";
+    public static final NamedAnalyzer WHITESPACE_ANALYZER = new NamedAnalyzer(
+        "whitespace", AnalyzerScope.INDEX, new WhitespaceAnalyzer());
```
That makes sense -- I'll move the construction of this analyzer to `JsonFieldMapper.Builder#build` to match what we do in `KeywordFieldMapper`.

jtibshirani commented on docs/reference/mapping/types/embedded-json.asciidoc:
```diff
@@ -0,0 +1,212 @@
+keywords. When sorting, this implies that values are compared lexicographically.
+
+Finally, because of the way leaf values are stored in the index, the null
+character `\0` is not allowed to appear in the keys of the JSON object.
```
Thanks, I had forgotten these weren't allowed by default.

jtibshirani committed:
Address code review feedback.

* Don't explicitly mention that '\0' is not allowed in keys.
* Use String#indexOf in FieldTypeLookup#fieldDepth.
* Construct the whitespace analyzer once per field mapper.

jtibshirani commented:
> However, this is not the case for stored fields, which makes me wonder whether we should leave it unsupported for now.

Looking at this again, I also find the behavior around stored fields to be a bit unintuitive:
- Only the root field is stored, it is not possible to load the keyed fields through `stored_fields`.
- We store the whole JSON input, which differs from what is indexed and stored in docvalues.
- We don't preserve the original formatting because we parse the JSON block, then reconstruct it to create the stored field.

I guess the main use case would be if a user wanted to retrieve the field in a search, but that using source filtering would be too expensive. I would be okay leaving it unsupported for now because the design is not that clean. @colings86 and @romseygeek checking since you reviewed this earlier -- what do you think about removing support for stored fields?

colings86 commented:
I’m fine with removing support for stored field until we have a more
intuitive way of exposing it

On Wed, 29 May 2019 at 19:20, Julie Tibshirani <notifications@github.com>
wrote:

> However, this is not the case for stored fields, which makes me wonder
> whether we should leave it unsupported for now.
>
> Looking at this again, I also find the behavior around stored fields to be
> a bit unintuitive:
>
>    - Only the root field is stored, it is not possible to load the keyed
>    fields through stored_fields.
>    - We store the whole JSON input, which differs from what is indexed
>    and stored in docvalues.
>    - We don't preserve the original formatting because we parse the JSON
>    block, then reconstruct it to create the stored field.
>
> I guess the main use case would be if a user wanted to retrieve the field
> in a search, but that using source filtering would be too expensive. I
> would be okay leaving it unsupported for now because the design is not that
> clean. @colings86 <https://github.com/colings86> and @romseygeek
> <https://github.com/romseygeek> checking since you reviewed this earlier
> -- what do you think about removing support for stored fields?
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/elastic/elasticsearch/pull/42541?email_source=notifications&email_token=AABZZO5HYMRANJE7AG6QDY3PX3CORA5CNFSM4HPR3JQ2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGODWQGWDY#issuecomment-497052431>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AABZZO5CGGGKL4LTYPPKZHLPX3CORANCNFSM4HPR3JQQ>
> .
>

jtibshirani commented on server/src/main/java/org/elasticsearch/index/mapper/FieldTypeLookup.java:
```diff
@@ -36,15 +38,24 @@
     final CopyOnWriteHashMap<String, MappedFieldType> fullNameToFieldType;
     private final CopyOnWriteHashMap<String, String> aliasToConcreteName;
 
+    private final CopyOnWriteHashMap<String, JsonFieldMapper> fullNameToJsonMapper;
```
I had actually tried this, but found it cleaner to use `JsonFieldMapper` here compared to the other option, where we use `RootJsonFieldType` and create the `KeyedJsonFieldType` objects using it. I found it nice that `JsonFieldMapper` contained consistent pairs of methods (`fieldType()` and `keyedFieldType(), `name()` and `keyedFieldName()`). To me it's acting in its role as 'field type provider'.

jtibshirani committed:
Remove support for stored fields.

jtibshirani committed:
Improve the field type documentation.

* Remove comment about saving space.
* Emphasize the similarity to keyword fields.
* Line wrap at 80 characters.

jtibshirani commented:
@jpountz this is ready for another look. I ended up force-pushing -- the first new commit since you reviewed is d1f8ad2.

jtibshirani committed:
Rename 'embedded JSON' to 'flattened object'.

The code refers to 'flat object' in some places for brevity.

jtibshirani committed:
Merge remote-tracking branch 'upstream/master' into object-fields

jtibshirani committed:
Fix full text queries, which were broken after the merge.

jtibshirani commented:
@jpountz @colings86 this is ready for another look. The last commit since you reviewed is 7237b2b0967bf710cc033b8f40e6e082227691c4 ('Remove the experimental tag.')

jpountz commented on server/src/main/java/org/elasticsearch/index/mapper/FlatObjectFieldParser.java:
```diff
@@ -0,0 +1,178 @@
+ * A helper class for {@link FlatObjectFieldMapper} parses a JSON object
+ * and produces a pair of indexable fields for each leaf value.
+ */
+public class FlatObjectFieldParser {
```
can it be pkg-private?

jtibshirani replied:
👍

jpountz commented on server/src/test/java/org/elasticsearch/search/query/SearchQueryIT.java:
```diff
@@ -1757,5 +1757,5 @@ public void testFieldAliasesForMetaFields() throws Exception {
         DocumentField field = hit.getFields().get("id-alias");
         assertThat(field.getValue().toString(), equalTo("1"));
-   }
+    }
```
maybe undo changes to this file since they are unrelated

jtibshirani replied:
👍

jpountz reviewed (approved):
Apart from minor comments, it looks good to me.

jtibshirani commented on docs/reference/mapping/types/flattened.asciidoc:
```diff
@@ -0,0 +1,191 @@
+    whitespace when building a query for this field. Accepts `true` or `false`
+    (default).
+
+<<mapping-store,`store`>>::
```
Oops, thanks I missed this.

jtibshirani commented on rest-api-spec/src/main/resources/rest-api-spec/test/search/160_exists_query.yml:
```diff
@@ -1328,3 +1328,57 @@ setup:
+        index:  flat_object_test
+        body:
+          mappings:
+            dynamic: false
```
👍 removed.

jtibshirani committed:
Address code review feedback and remove leftover references to 'json'.

jtibshirani commented:
@elasticmachine run elasticsearch-ci/default-distro 
@elasticmachine run elasticsearch-ci/bwc

jtibshirani committed:
Merge remote-tracking branch 'upstream/master' into object-fields

jtibshirani committed:
Merge remote-tracking branch 'upstream/master' into object-fields

jtibshirani committed:
Pull the flattened field type into a mapper plugin. (#43250)

This PR pulls `FlatObjectFieldMapper` into its own `MapperPlugin`. To do so it
introduces a new interface `DynamicKeyFieldMapper` with the method
`keyedFieldType(String key)`, which gives the opportunity to return a special
field type for a subfield.

jtibshirani committed:
Merge remote-tracking branch 'upstream/master' into object-fields

jtibshirani committed:
License the flattened mapper as basic. (#43690)

This PR pulls the `flattened` mapper plugin into the xpack directory as its own
feature.

jtibshirani committed:
Merge remote-tracking branch 'upstream/master' into object-fields

jtibshirani committed:
Merge remote-tracking branch 'upstream/master' into object-fields
